pkg/innerring/node.go

```diff
@@ -0,0 +1,354 @@
+//   - cheque ID set to the specified withdrawal transaction ID
+//   - lock account to the first 20 bytes of the specified withdrawal transaction ID
+//   - user account and assets' amount as it is
+//
+// If local node is configured with non-separated Mainchain, NeoFS contract is
+// expected to be deployed in the NeoFS Sidechain.
+func (x *node) ApproveWithdrawal(userAcc util.Uint160, amount uint64, withdrawTx util.Uint256) error {
+	if x.mainChain.enabled {
+		var lockAcc util.Uint160
+		copy(lockAcc[:], withdrawTx[:])
+
+		var prm neofscontract.ChequePrm
+		prm.SetID(withdrawTx.BytesBE())
+		prm.SetAmount(int64(amount))
+		prm.SetUser(userAcc)
+		prm.SetLock(lockAcc)
+		prm.SetHash(withdrawTx)
+
+		err := x.mainChain.neoFSClient.Cheque(prm)
+		if err != nil {
+			return fmt.Errorf("call cheque method of NeoFS contract client: %w", err)
+		}
+
+		return nil
+	}
+
+	const method = "cheque"
+	contract := x.contracts.get(contractNeoFS)
+
+	if !x.sidechain.NotaryServiceEnabled() {
+		err := x.sidechain.CallContractMethod(contract, method, withdrawTx, userAcc, amount, withdrawTx)
+		if err != nil {
+			return fmt.Errorf("call '%s' method of the NeoFS contract: %w", method, err)
+		}
+
+		return nil
+	}
+
+	alphabetKeys, err := x.sidechain.Committee()
+	if err != nil {
+		return err
+	}
+
+	m := smartcontract.GetDefaultHonestNodeCount(len(alphabetKeys))
+
+	multiSigScript, err := smartcontract.CreateMultiSigRedeemScript(m, alphabetKeys)
+	if err != nil {
+		return fmt.Errorf("create multi-sig redeem script for Alphabet: %w", err)
+	}
+
+	var tx transaction.Transaction
+
+	// main net: &transaction.Signer{Scopes: transaction.CalledByEntry}
+	// morph: &transaction.Signer{	Scopes: transaction.Global
+
+	tx.Signers = []transaction.Signer{
+		{
+			Account: x.contracts.get(contractProxy),
+			Scopes:  transaction.None,
+		},
+		{
+			Account: hash.Hash160(multiSigScript),
+			Scopes:  transaction.CalledByEntry, // transaction.Global for sidechain
+		},
+		{
+			Account: notary.Hash, // notary contract
+			Scopes:  transaction.None,
+		},
+	}
+
+	callRes, err := x.sidechain.TestCallContractMethod(contract, method, withdrawTx, userAcc, amount, withdrawTx)
+	if err != nil {
+		return fmt.Errorf("test call '%s' method of the NeoFS contract: %w", method, err)
+	}
+
+	alphabetMultiSigAcc, err := x.buildAlphabetMultiSigAccount(alphabetKeys)
+	if err != nil {
+		return fmt.Errorf("build alphabet multi-sig account: %w", err)
+	}
+
+	withdrawHeight, err := x.sidechain.TransactionBlock(withdrawTx)
+	if err != nil {
+		return fmt.Errorf("get withdrawal transaction block height: %w", err)
+	}
+
+	tx.SystemFee = callRes.ConsumedGAS
+	tx.Nonce = binary.BigEndian.Uint32(withdrawTx[:4])
+	tx.ValidUntilBlock = withdrawHeight + 50
+	tx.Script = callRes.Script
+	tx.Attributes = []transaction.Attribute{{
+		Type: transaction.NotaryAssistedT,
+		Value: &transaction.NotaryAssisted{
+			NKeys: uint8(len(alphabetKeys)),
+		},
+	}}
+	tx.Scripts = []transaction.Witness{
```

Comment:
Unsatisfied with the notary actor interface?

Reply:
I'll try.

Reply:
Done in https://github.com/nspcc-dev/neofs-node/pull/2211/commits/9709c1e6ccd1c2182dbbb79676ab4d4cf2a41d09. I couldn't test changes in the NeoFS Dev Env because I didn't succeeded in withdraw ops without Mainchain. Are they even real without the Mainchain?